pkg/datapath/loader/netlink.go

```diff
@@ -66,60 +66,123 @@ func replaceQdisc(ifName string) error {
+		WithField("ifindex", link.Attrs().Index)
+
+	// Load the ELF from disk.
+	l.Debug("Loading CollectionSpec from ELF")
+	spec, err := bpf.LoadCollectionSpec(objPath)
+	if err != nil {
+		return nil, fmt.Errorf("loading eBPF ELF: %w", err)
+	}
+
+	if spec.Programs[progName] == nil {
+		return nil, fmt.Errorf("no program %s found in eBPF ELF", progName)
 	}
 
-	// FIXME: replace exec with native call
-	if xdp {
-		loaderProg = "ip"
-		args = []string{"-force", "link", "set", "dev", ifName, xdpMode,
-			"obj", objPath, "sec", progSec}
-	} else {
-		loaderProg = "tc"
+	// Load the CollectionSpec into the kernel, picking up any pinned maps from
+	// bpffs in the process.
+	finalize := func() {}
+	opts := ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{PinPath: bpf.MapPrefixPath()},
+	}
+	l.Debug("Loading Collection into kernel")
+	coll, err := bpf.LoadCollection(spec, opts)
+	if errors.Is(err, ebpf.ErrMapIncompatible) {
+		// Temporarily rename bpffs pins of maps whose definitions have changed in
+		// a new version of a datapath ELF.
+		l.Debug("Starting bpffs map migration")
+		if err := bpf.StartBPFFSMigration(bpf.MapPrefixPath(), spec); err != nil {
+			return nil, fmt.Errorf("Failed to start bpffs map migration: %w", err)
+		}
 
-		tcPrio := strconv.Itoa(option.Config.TCFilterPriority)
-		log.Debugf("tc filter using priority %s for interface %s", tcPrio, ifName)
-		args = []string{"filter", "replace", "dev", ifName, progDirection,
-			"prio", tcPrio, "handle", "1", "bpf", "da", "obj", objPath,
-			"sec", progSec,
+		finalize = func() {
+			l.Debug("Finalizing bpffs map migration")
+			if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), spec, false); err != nil {
+				l.WithError(err).Error("Could not finalize bpffs map migration")
+			}
 		}
+
+		// Retry loading the Collection after starting map migration.
+		l.Debug("Retrying loading Collection into kernel after map migration")
+		coll, err = bpf.LoadCollection(spec, opts)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error loading eBPF collection into the kernel: %w", err)
 	}
+	defer coll.Close()
 
-	// If the iproute2 call below is successful, any 'pending' map pins will be removed.
-	// If not, any pending maps will be re-pinned back to their initial paths.
-	cmd := exec.CommandContext(ctx, loaderProg, args...).WithFilters(libbpfFixupMsg)
-	if _, err := cmd.CombinedOutput(log, true); err != nil {
-		// Program/object replacement unsuccessful, revert bpffs migration.
-		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), objPath, true); err != nil {
+	// Avoid attaching a prog to a stale interface.
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	l.Debug("Attaching program to interface")
+	if err := attachProgram(link, coll.Programs[progName], directionToParent(direction), xdpModeToFlag(xdpMode)); err != nil {
+		// Program replacement unsuccessful, revert bpffs migration.
+		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), spec, true); err != nil {
```

Comment:
Looks like `err` is still being shadowed here, is that intentional?

https://github.com/cilium/cilium/pull/19159/files#r887403069

Reply:
Ah, thought you wanted to leave this as-is. We'll log the migration error and return the netlink one. The failed migration isn't actually the error we're trying to recover from.

Reply:
Ah yes this is fine sorry. I couldn't quite tell what the resolution was from the previous time.